File: imp/java/src/fit/Parse.java
package fit;

// Copyright (c) 2002 Cunningham & Cunningham, Inc.
// Released under the terms of the GNU General Public License version 2 or later.

import java.io.*;
import java.text.ParseException;

public class Parse {

    public String leader;
    public String tag;
    public String body;
    public String end;
    public String trailer;

    public Parse more;
    public Parse parts;

    public Parse (String tag, String body, Parse parts, Parse more) {
        this.leader = "\n";
        this.tag = "<"+tag+">";
        this.body = body;
        this.end = "</"+tag+">";
        this.trailer = "";
        this.parts = parts;
        this.more = more;
    }

    public static String tags[] = {"table", "tr", "td"};

    public Parse (String text) throws ParseException {
        this (text, tags, 0, 0);
    }

    public Parse (String text, String tags[]) throws ParseException {
        this (text, tags, 0, 0);
    }

    public Parse (String text, String tags[], int level, int offset) throws ParseException {
        String lc = text.toLowerCase();
        int startTag = lc.indexOf("<"+tags[level]);
        int endTag = lc.indexOf(">", startTag) + 1;
//        int startEnd = lc.indexOf("</"+tags[level], endTag);
		int startEnd = findMatchingEndTag(lc, endTag, tags[level], offset);
        int endEnd = lc.indexOf(">", startEnd) + 1;
        int startMore = lc.indexOf("<"+tags[level], endEnd);
        if (startTag<0 || endTag<0 || startEnd<0 || endEnd<0) {
            throw new ParseException ("Can't find tag: "+tags[level], offset);
        }

        leader = text.substring(0,startTag);
        tag = text.substring(startTag, endTag);
        body = text.substring(endTag, startEnd);
        end = text.substring(startEnd,endEnd);
        trailer = text.substring(endEnd);

        if (level+1 < tags.length) {
            parts = new Parse (body, tags, level+1, offset+endTag);
            body = null;
        }
		else { // Check for nested table
			int index = body.indexOf("<" + tags[0]);
			if (index >= 0) {
				parts = new Parse(body, tags, 0, offset + endTag);
				body = "";
			}
		}

        if (startMore>=0) {
            more = new Parse (trailer, tags, level, offset+endEnd);
            trailer = null;
        }
    }

	/* Added by Rick Mugridge, Feb 2005 */
	protected static int findMatchingEndTag(String lc, int matchFromHere, String tag, int offset) throws ParseException {
		int fromHere = matchFromHere;
		int count = 1;
		int startEnd = 0;
		while (count > 0) {
			int embeddedTag = lc.indexOf("<" + tag, fromHere);
			int embeddedTagEnd = lc.indexOf("</" + tag, fromHere);
			// Which one is closer?
			if (embeddedTag < 0 && embeddedTagEnd < 0)
				throw new ParseException("Can't find tag: " + tag, offset);
			if (embeddedTag < 0)
				embeddedTag = Integer.MAX_VALUE;
			if (embeddedTagEnd < 0)
				embeddedTagEnd = Integer.MAX_VALUE;
			if (embeddedTag < embeddedTagEnd) {
				count++;
				startEnd = embeddedTag;
				fromHere = lc.indexOf(">", embeddedTag) + 1;
			}
			else if (embeddedTagEnd < embeddedTag) {
				count--;
				startEnd = embeddedTagEnd;
				fromHere = lc.indexOf(">", embeddedTagEnd) + 1;
			}
		}
		return startEnd;
	}

    public int size() {
        return more==null ? 1 : more.size()+1;
    }

    public Parse last() {
        return more==null ? this : more.last();
    }

    public Parse leaf() {
        return parts==null ? this : parts.leaf();
    }

    public Parse at(int i) {
        return i==0 || more==null ? this : more.at(i-1);
    }

    public Parse at(int i, int j) {
        return at(i).parts.at(j);
    }

    public Parse at(int i, int j, int k) {
        return at(i,j).parts.at(k);
    }

    public String text() {
    	return htmlToText(body);
    }
    
    public static String htmlToText(String s)
    {
		s = normalizeLineBreaks(s); 
    	s = removeNonBreakTags(s);
		s = condenseWhitespace(s);
		s = unescape(s);
    	return s;
    }

    private static String removeNonBreakTags(String s) {
        int i=0, j;
        while ((i=s.indexOf('<',i))>=0) {
            if ((j=s.indexOf('>',i+1))>0) {
                if (!(s.substring(i, j+1).equals("<br />"))) {
                	s = s.substring(0,i) + s.substring(j+1);
                } else i++;
            } else break;
        }
        return s;
    }

    public static String unescape(String s) {
    	s = s.replaceAll("<br />", "\n");
		s = unescapeEntities(s);
		s = unescapeSmartQuotes(s);
        return s;
    }

	private static String unescapeSmartQuotes(String s) {
		s = s.replace('\u201c', '"');
		s = s.replace('\u201d', '"');
		s = s.replace('\u2018', '\'');
		s = s.replace('\u2019', '\''); 
		return s;
	}

	private static String unescapeEntities(String s) {
		s = s.replaceAll("&lt;", "<");
		s = s.replaceAll("&gt;", ">");
		s = s.replaceAll("&nbsp;", " ");
		s = s.replaceAll("&quot;", "\"");
		s = s.replaceAll("&amp;", "&");
		return s;
	}

	private static String normalizeLineBreaks(String s) {
		s = s.replaceAll("<\\s*br\\s*/?\\s*>", "<br />");
		s = s.replaceAll("<\\s*/\\s*p\\s*>\\s*<\\s*p.*?>", "<br />");
		return s;
	}
	
    public static String condenseWhitespace(String s) {
    	final char NON_BREAKING_SPACE = (char)160;
    	
    	s = s.replaceAll("\\s+", " ");
		s = s.replace(NON_BREAKING_SPACE, ' ');
		s = s.replaceAll("&nbsp;", " ");
		s = s.trim();
    	return s;    	
    }

    public void addToTag(String text) {
        int last = tag.length()-1;
        tag = tag.substring(0,last) + text + ">";
    }

    public void addToBody(String text) {
        body = body + text;
    }

    public void print(PrintWriter out) {
        out.print(leader);
        out.print(tag);
        if (parts != null) {
            parts.print(out);
        } else {
            out.print(body);
        }
        out.print(end);
        if (more != null) {
            more.print(out);
        } else {
            out.print(trailer);
        }
    }

    public static int footnoteFiles=0;
    public String footnote () {
        if (footnoteFiles>=25) {
            return "[-]";
        } else {
            try {
                int thisFootnote = ++footnoteFiles;
                String html = "footnotes/" + thisFootnote + ".html";
                File file = new File("Reports/" + html);
                file.delete();
                PrintWriter output = new PrintWriter(new BufferedWriter(new FileWriter(file)));
                print(output);
                output.close();
                return "<a href=/fit/Release/Reports/" + html + "> [" + thisFootnote + "]</a>";
            } catch (IOException e) {
                return "[!]";
            }
        }
    }
}

File: imp/ruby/lib/fit/parse.rb
# Copyright (c) 2002 Cunningham & Cunningham, Inc.
# Released under the terms of the GNU General Public License version 2 or later.

module Fit

  class ParseException < Exception
    attr_reader :error_offset
    def initialize message, offset
      super message
      @error_offset = offset
    end
  end

  # This ParseHolder class is necessary to cope with the different constructors
  # in the Java version of FIT. One of them is the constructor for the Parse Ruby
  # class, the other is a static method in the ParseHolder class.
  #
  # ParseHolder also maintain all the methods belonging to Parse which are instead
  # called whenever in FIT a ParseHolder is created instead of a Parse.
  class ParseHolder

    attr_accessor :leader, :tag, :body, :end, :trailer
    attr_accessor :more, :parts

    def ParseHolder.create tag, body, parts, more
      p = new
      p.leader = "\n"
      p.tag = "<#{tag}>"
      p.body = body
      p.end = "</#{tag}>"
      p.trailer = ""
      p.parts = parts
      p.more = more
      p
    end

    def leaf
      @parts.nil? ? self : @parts.leaf
    end

    def add_to_tag text
      @tag = @tag[0..-2] + " #{text}>"
    end

    def add_to_body text
      @body += text
    end
    
    def at i, *rest
      node = (i == 0 || @more.nil?) ? self : @more.at(i - 1)
      rest.each do |j|
        node = node.parts.at(j)
      end
      node
    end

    def text
      Parse.html_to_text @body
    end

    def last
      @more.nil? ? self : @more.last
    end

    def size
      @more.nil? ? 1 : @more.size + 1
    end

    def print out
      out.print @leader
      out.print @tag
      if @parts
        @parts.print out
      else
        out.print @body
      end
      out.print @end
      if @more
        @more.print out
      else
        out.print @trailer
      end
    end

  end

  class Parse < ParseHolder

    DEFAULT_TAGS = ['table', 'tr', 'td']

    @@footnote_files = 0

    def initialize text, tags = DEFAULT_TAGS, level = 0, offset = 0
      tag = tags[level]
      raise ParseException.new("Can't find tag: #{tag}", offset) unless text =~ /(<#{tag}.*?>)(.*?)(<\/#{tag}.*?>)/im
      @leader, @tag, @body, @end, @trailer = $`, $1, $2, $3, $'
      match = $&

      if level + 1 < tags.size
        @parts = Parse.new @body, tags, level + 1, offset + @leader.size + @tag.size
        @body = nil
      end

      if @trailer =~ /<#{tag}/i
        @more = Parse.new @trailer, tags, level, offset + match.size
        @trailer = nil
      end
    end

    def Parse.html_to_text s
      str = s.gsub(%r{<\s*br\s*/?\s*>}, '<br />').gsub(%r{<\s*/\s*p\s*>\s*<\s*p\s*>}, '<br />')
      unescape(condense_whitespace(remove_tags(str)))
    end

    def Parse.remove_tags s
      s.gsub(/<.*?>/m) { $& == '<br />' ? $& : '' }
    end

    def Parse.unescape s
      str = s.gsub %r{<br />}, "\n"
      # unescape HTML entities
      str = str.gsub(%r{&lt;}, '<').gsub(%r{&gt;}, '>').gsub(%r{&nbsp;}, ' ').gsub(%r{&quot;}, '"').gsub(%r{&amp;}, '&')
      # unescape smart quotes
      str.gsub(/\223/, '"').gsub(/\224/, '"').gsub(/\221/, "'").gsub(/\222/, "'")
    end

    def Parse.condense_whitespace s
      # NOT_BREAKING_SPACE is decimal character 160, hex a0, oct 240
      s.gsub(%r{\s+}, ' ').gsub(/\240/, ' ').gsub(%r{&nbsp;}, ' ').strip
    end

    def footnote
      return '[-]' if @@footnote_files >= 25
      begin
        this_footnote = (@@footnote_files += 1)
        html = "footnotes/#{this_footnote}.html"
        f = File.new("reports/#{html}", 'w')
        print f
        f.close
        return "<a href=/fit/Release/Reports/#{html}>[#{this_footnote}]</a>"
      rescue Exception => e
        puts e.message
        puts e.backtrace
        return '[!]'
      end
    end

  end

end
